### PR TITLE
Endpoint for smoothing data #612

### LINKED ISSUE #612: Endpoint for smoothing data
subtask for #606 

add endpoint that return retrieved smoothed data from analytic

### PR DESCRIPTION
**Pull request by rozetko**
Closes #627 
Closes #612 

changes:
- new task type for data processing
- `ProcessingDetector` base class for classes that can process data
- `GET /query`: return smoothed dataset instead of raw for anomaly analytic units
- move detection result concatenation to Detectors
- send proper payload with anomaly detection task

### CODE REVIEW

File: analytics/analytics/detectors/anomaly_detector.py [hunk 1]
```diff
@@ -48,13 +50,34 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
...
         last_detection_time = utils.convert_pd_timestamp_to_ms(last_dataframe_time)
+        # TODO: ['lastValue'] -> .last_value
+        cache['lastValue'] = smoothed_data[-1]
+
         return DetectionResult(cache, segments, last_detection_time)
 
     def consume_data(self, data: pd.DataFrame, cache: Optional[ModelCache]) -> Optional[DetectionResult]:
-        self.detect(data, cache)
+        if cache is None:
+            msg = f'consume_data get invalid cache {cache} for task {self.analytic_unit_id}'
```
Comment (jonyrock) [inline analytics/analytics/detectors/anomaly_detector.py]: got

File: server/src/routes/data_router.ts [hunk 2]
```diff
@@ -47,8 +50,32 @@ async function query(ctx: Router.IRouterContext) {
...
+    data,
+    analyticUnitType,
+    detector,
+    cache
+  };
+
+  const processingTask = new AnalyticsTask(analyticUnitId, AnalyticsTaskType.PROCESS, payload);
+  let result = await AnalyticsController.runTask(processingTask);
+  if(result.status !== AnalyticUnit.AnalyticUnitStatus.SUCCESS) {
+    throw new Error(result.error);
```
Comment (jonyrock) [inline server/src/routes/data_router.ts]: please wrap inner error to your level of abstraction

File: analytics/analytics/detectors/anomaly_detector.py [hunk 3]
```diff
@@ -48,13 +50,34 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
...
         last_detection_time = utils.convert_pd_timestamp_to_ms(last_dataframe_time)
+        # TODO: ['lastValue'] -> .last_value
+        cache['lastValue'] = smoothed_data.values[-1]
+
         return DetectionResult(cache, segments, last_detection_time)
 
     def consume_data(self, data: pd.DataFrame, cache: Optional[ModelCache]) -> Optional[DetectionResult]:
-        self.detect(data, cache)
+        if cache is None:
+            msg = f'consume_data get invalid cache {cache} for task {self.analytic_unit_id}'
```
Comment (amper43) [inline analytics/analytics/detectors/anomaly_detector.py]:
```suggestion
            msg = f'consume_data got invalid cache {cache} for task {self.analytic_unit_id}'
```

File: analytics/analytics/detectors/anomaly_detector.py [hunk 4]
```diff
@@ -1,22 +1,24 @@
...
 MAX_DEPENDENCY_LEVEL = 100
 MIN_DEPENDENCY_FACTOR = 0.1
 logger = logging.getLogger('ANOMALY_DETECTOR')
 
 
-class AnomalyDetector(Detector):
+class AnomalyDetector(ProcessingDetector):
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, analytic_unit_id):
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]:
```suggestion
    def __init__(self, analytic_unit_id: AnalyticUnitId):
```

File: server/src/controllers/analytics_controller.ts [hunk 5]
```diff
@@ -558,3 +573,49 @@ async function runDetectionOnExtendedSpan(
...
+export async function getHSR(analyticUnit: AnalyticUnit.AnalyticUnit, from: number, to: number) {
+
+  try {
+    const grafanaUrl = getGrafanaUrl(analyticUnit.grafanaUrl);
+    const data = await queryByMetric(analyticUnit.metric, grafanaUrl, from, to, HASTIC_API_KEY);
+
+    if(analyticUnit.detectorType !== AnalyticUnit.DetectorType.ANOMALY) {
+      return data;
+    } else {
+
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
```

File: server/src/models/analytic_units/index.ts [hunk 6]
```diff
@@ -3,7 +3,8 @@ import { AnalyticUnitId, AnalyticUnitStatus, DetectorType, ANALYTIC_UNIT_TYPES }
 import { AnalyticUnit } from './analytic_unit_model';
 import { PatternAnalyticUnit } from './pattern_analytic_unit_model';
 import { ThresholdAnalyticUnit } from './threshold_analytic_unit_model';
-import {
+import { AnomalyAnalyticUnit } from './anomaly_analytic_unit_model';
+ import {
```
Comment (rozetko) [inline server/src/models/analytic_units/index.ts]:
```suggestion
import {
```

File: server/src/controllers/analytics_controller.ts [hunk 7]
```diff
@@ -558,3 +573,48 @@ async function runDetectionOnExtendedSpan(
...
+        data: data.values,
+        analyticUnitType,
+        detector,
+        cache
+      };
+
+      const processingTask = new AnalyticsTask(analyticUnit.id, AnalyticsTaskType.PROCESS, payload);
+      const result = await runTask(processingTask);
+      if(result.status !== AnalyticUnit.AnalyticUnitStatus.SUCCESS) {
+        throw new Error(`got error while procesing data ${result.error}`);
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
        throw new Error(`Data processing error: ${result.error}`);
```

File: server/src/controllers/analytics_controller.ts [hunk 8]
```diff
@@ -558,3 +573,48 @@ async function runDetectionOnExtendedSpan(
...
+        throw new Error(`Data processing error: ${result.error}`);
+      }
+      return { values: result.payload.data, columns: data.columns }
+    }
+  } catch (err) {
+    const message = err.message || JSON.stringify(err);
+    await AnalyticUnit.setStatus(analyticUnit.id, AnalyticUnit.AnalyticUnitStatus.FAILED, message);
+    throw new Error(message);
+  }
+
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
```

File: server/src/controllers/analytics_controller.ts [hunk 9]
```diff
@@ -215,23 +219,34 @@ export async function runLearning(id: AnalyticUnit.AnalyticUnitId, from?: number
...
+        taskPayload.threshold = {
+          value: (analyticUnit as ThresholdAnalyticUnit).value,
+          condition: (analyticUnit as ThresholdAnalyticUnit).condition
+        };
+        break;
+      case AnalyticUnit.DetectorType.ANOMALY:
+        taskPayload.anomaly = {
+          alpha: (analyticUnit as AnomalyAnalyticUnit).alpha,
+          confidence: (analyticUnit as AnomalyAnalyticUnit).confidence
+        }
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
        };
```

File: server/src/controllers/analytics_controller.ts [hunk 10]
```diff
@@ -215,23 +219,34 @@ export async function runLearning(id: AnalyticUnit.AnalyticUnitId, from?: number
...
-      taskPayload.segments = segmentObjs;
-    } else if(detector === AnalyticUnit.DetectorType.THRESHOLD) {
-      taskPayload.threshold = {
-        value: (analyticUnit as ThresholdAnalyticUnit).value,
-        condition: (analyticUnit as ThresholdAnalyticUnit).condition
-      };
+    switch(detector) {
+      case AnalyticUnit.DetectorType.PATTERN:
+        let segments = await Segment.findMany(id, { labeled: true });
+        if (segments.length === 0) {
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
        if(segments.length === 0) {
```

File: server/src/controllers/analytics_controller.ts [hunk 11]
```diff
@@ -558,3 +573,47 @@ async function runDetectionOnExtendedSpan(
   await Detection.insertSpan(detection);
   return detection;
 }
+
+export async function getHSR(analyticUnit: AnalyticUnit.AnalyticUnit, from: number, to: number) {
+
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
```

**Pull request merged**